exercise/exercisecollection_models.py

```diff
@@ -0,0 +1,212 @@
+            current_submission.clean()
+            current_submission.save()
+        else:
+            submissions = self.get_submissions_for_student(user.userprofile)
+            current_submission = submissions[0]
+
+
+        new_grade = self.get_points(user)
+
+        if new_grade is None:
+            return
+
+        if new_grade == current_submission.grade and not forced:
+            return
+
+
+        grading_data, feedback = self._generate_grading_data(user.userprofile)
+
+
+        current_submission.grade = new_grade
+        current_submission.submission_time = timezone.now()
+        current_submission.status = Submission.STATUS.READY
+        current_submission.submitters = [user.userprofile]
+        current_submission.grading_data = grading_data
+        current_submission.feedback = feedback
+        current_submission.clean()
+        current_submission.save()
+
+
+
+    # Check if user has a submission for this exercise
+    def is_submitted(self, user):
+        return self.get_submissions_for_student(user.userprofile).count() > 0
+
+
+    # Property to access max_points in target category
+    @property
+    def target_category_maxpoints(self):
+        max_points = 0
+
+        for exercise in self.exercises:
+            max_points += exercise.max_points
+
+        return max_points
+
+
+    # Property to access exercises in target category
+    @property
+    def exercises(self):
+        return BaseExercise.objects.filter(
+            category=self.target_category
+        )
+
+
+    # Generates feedback and grading_data
+    # Feedback is in HTML format
+    # grading_data is currently blank
+    def _generate_grading_data(self, profile):
+        feedback = ""
+        grading_data = ""
+
+        exercise_counter = 1
+        for exercise in self.exercises:
+
+            submission = UserExerciseSummary(exercise, profile.user).best_submission
+            if submission is None:
+                grade = 0
+            else:
+                grade = submission.grade
+
+            feedback += "Exercise {}: {}/{}\n  Course: {} - {}\n  Exercise: {}\n".format(
+                exercise_counter,
+                grade,
+                exercise.max_points,
+                exercise.category.course_instance.course.name,
+                exercise.category.course_instance.instance_name,
+                exercise.name,
+            )
+            exercise_counter += 1
+
+        feedback = "<pre>\n" + feedback + "\n</pre>\n"
+        return {"grading_data": grading_data}, feedback
+
+    def get_load_url(self, language, request, students, url_name="exercise"):
+        pass
+
+
+# Updates submissions if new submission is in any ExerciseCollection's target category.
+# ! Probably needs Cache-optimization
+@receiver(post_save, sender=Submission)
+def update_exercise_collection_submission(sender, instance, **kwargs):
+    collections = ExerciseCollection.objects.filter(target_category=instance.exercise.category)
```

Comment:
This should create only max two db queries when there is no exercisecollection. One to check that out and possibly to resolve exercise.

That should be fine.